### ISSUE TITLE
[css-motion] Grammar corrections

### ISSUE THREAD

**Issue by shans** [w3c/fxtf-drafts#42]
(with @tabatkins)

```
offset-path: <url> | <angle> && <size>? && contain? | 
             <basic-shape> && <geometry-box>? | <path()> | none
```

i.e., you can have one of the following:
- a url
- an angle, optionally with a size and optionally with the contain keyword, in any order
- a basic shape, optionally with a geometry box, in any order
- a path
- none 

```
offset-rotation: [ auto | reverse ] || <angle>
```

i.e. you can have auto/reverse, OR an angle, OR both, in any order

offset: There are multiple problems here. In addition to the problem that Eric identified (offset-position and offset-anchor are both positions)
- offset-distance is ambiguous with a position when it appears next to it (so offset-distance could not appear after offset-position or offset-anchor); and
- multiple properties have the value auto.

Basically, it's very unlikely that this could ever work. We propose dropping the shorthand back to the original proposal (offset-path, offset-distance, offset-rotation). These are unambiguous with respect to each other, and likely to be the most commonly used parts of the specification. In other words:

```
offset: <offset-path> || <offset-distance> || <offset-rotation>
```

(Note double bar instead of double ampersand - this allows one or more of these properties to be specified whereas double ampersand requires all of them.)

**Comment by shans** [comment]
(btw happy to send a pull request if this all sounds good)

**Comment by ericwilligers** [comment]
The following worked well when offset-path couldn't be an angle:
 `offset: <offset-path> || <offset-distance> || <offset-rotation>`

But now offset-path and offset-rotation can both be angles, so the grammar correction doesn't solve the last of the following: 
- offset-distance is ambiguous with a position 
- multiple properties have the value auto
- multiple properties have position
- consecutive positions are ambiguous
- multiple properties have angle

One way to repair this would be to make the size non-optional in angle paths. Another would be to require property ordering in the shorthand.

**Comment by shans** [comment]
I'm starting to think we should specify separate shorthands for the motion-on-a-path and the distribution-around-a-point use cases.

**Comment by ewilligers** [comment]
By fixing the order and making offset-distance compulsory, we can avoid problems:

`<offset-path>? <offset-distance> <offset-rotation>? <offset-position>? [/ <offset-anchor>]?`

Note that any 'auto' ambiguities between offset-rotation and offset-position are harmless: it doesn't matter which property is being set explicitly and which is being set implicitly.

**Comment by shans** [comment]
Alternatively we can provide syntax support by preventing naked angles from appearing in <offset-path>:

```
<url> | ray(<angle> [<size> || contain]?) | [ <basic-shape> | <path()> ] || <geometry-box> | none
```

This is a better approach IMHO as it reduces the constraints in the shorthand. Shorthand constraints can be confusing to authors and make shorthands relatively more brittle.

**Comment by shans** [comment]
Sorry, copied the incorrect syntax from the specification. That should read:

```
<url> | ray(<angle> [<size> || contain]?) | <basic-shape> && <geometry-box>? | <path()> | none
```

The && syntax version that allows the angle to appear at any point in the ray function would probably be fine too:

```
<url> | ray(<angle> && <size>? && contain?) | <basic-shape> && <geometry-box>? | <path()> | none
```

**Comment by jihyerish** [comment]
Is `ray()` the new function for making the path with < angle > ?
Could we discuss this issue at TPAC with shorthand `offset`?

**Comment by shans** [comment]
Yes, by introducing a function we can make the shorthand a lot easier to use. Let's definitely discuss at TPAC.

**Comment by jihyerish** [comment]
At TPAC, the syntax of the shorthand `offset` is resolved to 

> < offset-position >? [ < offset-path > [ < offset-rotation > || < offset-distance > ] [ / < offset-anchor > ]? ]

  -  Related Log: https://logs.csswg.org/irc.w3.org/css/2016-09-20/#e724234

But this still has a problem.
Both `offset-position` and `offset-rotation` can have `auto`, there is still ambiguous between them.
For example, 
> offset: auto  90deg

Which would be the value for `offset-rotation`? `auto 90deg` or `90deg`?

This problem can be solved in 2 ways:

1) Add extra "/" between `offset-position` and `offset-path`
> < offset-position >? [ / < offset-path > [ < offset-rotation > || < offset-distance > ] [ / < offset-anchor > ]? ]

2) 
> [  <offset-position > | [ < offset-position > ? < offset-path > [ < offset-distance > ? && < offset-rotation > ? ]]] [/ < offset-anchor > ]?

**Comment by ewilligers** [comment]
With (2), perhaps Markdown swallowed the initial \<offset-position\>

`[ <offset-position> | [ <offset-position>? <offset-path> [ <offset-distance>? && <offset-rotation>? ]]] [ / <offset-anchor> ]?`



3)
`<offset-position> <offset-rotation>? [ / <offset-anchor> ]? |`
`<offset-position>? <offset-path> [ <offset-distance>? && <offset-rotation>? ] [ / <offset-anchor> ]?`

As in #47, I assume that for offset properties to have any effect,  \<offset-position\> and/or \<offset-path\> must be set.  \<offset-distance\> is meaningless without \<offset-path\>, but \<offset-rotation\> combined with \<offset-position\> can still be useful.
